# Exploratory Data Analysis

## Imports

In [1]:
%matplotlib inline
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

## Load the dataset

In [2]:
# Load the dataset

file_path = '../data/term-deposit-marketing-2020.csv'
df = pd.read_csv(file_path)

print('Data size',df.shape)
df.head()

Data size (40000, 14)


age           job  marital  education default  balance housing loan  \
0   58    management  married   tertiary      no     2143     yes   no   
1   44    technician   single  secondary      no       29     yes   no   
2   33  entrepreneur  married  secondary      no        2     yes  yes   
3   47   blue-collar  married    unknown      no     1506     yes   no   
4   33       unknown   single    unknown      no        1      no   no   

   contact  day month  duration  campaign   y  
0  unknown    5   may       261         1  no  
1  unknown    5   may       151         1  no  
2  unknown    5   may        76         1  no  
3  unknown    5   may        92         1  no  
4  unknown    5   may       198         1  no

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 40000 entries, 0 to 39999
Data columns (total 14 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   age        40000 non-null  int64 
 1   job        40000 non-null  object
 2   marital    40000 non-null  object
 3   education  40000 non-null  object
 4   default    40000 non-null  object
 5   balance    40000 non-null  int64 
 6   housing    40000 non-null  object
 7   loan       40000 non-null  object
 8   contact    40000 non-null  object
 9   day        40000 non-null  int64 
 10  month      40000 non-null  object
 11  duration   40000 non-null  int64 
 12  campaign   40000 non-null  int64 
 13  y          40000 non-null  object
dtypes: int64(5), object(9)
memory usage: 4.3+ MB


### 1. Age

In [4]:
print("Originally encoded as: \n\n", df.age.unique())

Originally encoded as: 

 [58 44 33 47 35 28 42 43 41 29 53 57 51 45 60 56 32 25 40 39 52 46 36 49
 59 37 50 54 55 48 24 38 31 30 27 34 23 26 61 22 21 20 66 62 83 75 67 70
 65 68 64 69 72 71 19 76 85 63 90 82 73 74 78 80 94 79 77 86 95 81]


In [5]:
df.age.value_counts()

32    1909
31    1826
33    1770
34    1730
35    1721
      ... 
94       1
86       1
90       1
81       1
95       1
Name: age, Length: 70, dtype: int64

### 2. Job

In [6]:
print("Originally encoded as: \n\n", df.job.unique())

Originally encoded as: 

 ['management' 'technician' 'entrepreneur' 'blue-collar' 'unknown'
 'retired' 'admin' 'services' 'self-employed' 'unemployed' 'housemaid'
 'student']


In [7]:
df.job.value_counts()

blue-collar      9383
management       8166
technician       6852
admin            4483
services         3910
retired          1437
self-employed    1414
entrepreneur     1405
unemployed       1104
housemaid        1087
student           524
unknown           235
Name: job, dtype: int64

### 3. Marital

In [8]:
print("Originally encoded as: \n\n", df.marital.unique())

Originally encoded as: 

 ['married' 'single' 'divorced']


In [9]:
df.marital.value_counts()

married     24386
single      10889
divorced     4725
Name: marital, dtype: int64

### 4. Education

In [10]:
print("Originally encoded as: \n\n", df.education.unique())

Originally encoded as: 

 ['tertiary' 'secondary' 'unknown' 'primary']


In [11]:
df.education.value_counts()

secondary    20993
tertiary     11206
primary       6270
unknown       1531
Name: education, dtype: int64

### 5. Default

In [12]:
print("Originally encoded as: \n\n", df.default.unique())

Originally encoded as: 

 ['no' 'yes']


In [13]:
df["default"].replace(['no', 'yes'], 
                      [0, 1],
                      inplace = True)

In [14]:
df.default.dtype

dtype('int64')

In [15]:
df.default.value_counts()

0    39191
1      809
Name: default, dtype: int64

### 6. Balance

In [16]:
print("Originally encoded as: \n\n", df.balance.unique())

Originally encoded as: 

 [  2143     29      2 ...   7222   3402 102127]


In [17]:
df.balance.value_counts()

0       3209
1        179
2        141
4        126
3        124
        ... 
4720       1
4313       1
2675       1
4752       1
4667       1
Name: balance, Length: 6849, dtype: int64

### 7. Housing

In [18]:
print("Originally encoded as: \n\n", df.housing.unique())

Originally encoded as: 

 ['yes' 'no']


In [19]:
df["housing"].replace(['no', 'yes'], 
                      [0, 1],
                      inplace = True)

In [20]:
df.housing.dtype

dtype('int64')

In [21]:
df.housing.value_counts()

1    24031
0    15969
Name: housing, dtype: int64

### 8. Loan

In [22]:
print("Originally encoded as: \n\n", df.loan.unique())

Originally encoded as: 

 ['no' 'yes']


In [23]:
df["loan"].replace(['no', 'yes'], 
                      [0, 1],
                      inplace = True)

In [24]:
df.loan.dtype

dtype('int64')

In [25]:
df.loan.value_counts()

0    33070
1     6930
Name: loan, dtype: int64

### 9. Contact

In [26]:
print("Originally encoded as: \n\n", df.contact.unique())

Originally encoded as: 

 ['unknown' 'cellular' 'telephone']


In [27]:
df.contact.value_counts()

cellular     24914
unknown      12765
telephone     2321
Name: contact, dtype: int64

### 10. Day

In [28]:
print("Originally encoded as: \n\n", df.day.unique())

Originally encoded as: 

 [ 5  6  7  8  9 12 13 14 15 16 19 20 21 23 26 27 28 29 30  2  3  4 11 17
 18 24 25  1 10 22 31]


In [29]:
df.day.value_counts()

20    2655
18    2173
21    1914
17    1812
6     1778
5     1691
19    1650
28    1650
14    1636
8     1620
7     1612
29    1600
15    1468
30    1457
9     1374
13    1325
12    1289
11    1272
16    1215
4     1183
2     1109
27     974
26     895
3      874
23     823
22     773
25     695
31     591
10     364
24     352
1      176
Name: day, dtype: int64

### 11. Month

In [30]:
print("Originally encoded as: \n\n", df.month.unique())

Originally encoded as: 

 ['may' 'jun' 'jul' 'aug' 'oct' 'nov' 'dec' 'jan' 'feb' 'mar' 'apr']


In [31]:
df.month.value_counts()

may    13532
jul     6380
aug     5215
jun     4734
nov     3598
apr     2718
feb     2296
jan     1176
mar      258
oct       80
dec       13
Name: month, dtype: int64

### 12. Duration

In [32]:
print("Originally encoded as: \n\n", df.duration.unique())

Originally encoded as: 

 [ 261  151   76 ... 1880 1460 2219]


In [33]:
df.duration.value_counts()

124     172
90      166
114     162
89      160
119     159
       ... 
1606      1
2453      1
1971      1
1875      1
2015      1
Name: duration, Length: 1535, dtype: int64

### 13. Campaign 

In [34]:
print("Originally encoded as: \n\n", df.campaign.unique())

Originally encoded as: 

 [ 1  2  3  5  4  6  7  8  9 10 11 12 13 19 14 24 16 32 18 22 15 17 25 21
 43 51 63 41 26 28 55 50 38 23 20 29 31 37 30 46 27 58 33 35 34 36 39 44]


In [35]:
df.campaign.value_counts()

1     14718
2     11153
3      4970
4      3297
5      1642
6      1228
7       699
8       526
9       318
10      262
11      197
12      154
13      131
14       93
15       83
16       78
17       69
18       51
19       44
20       43
21       35
22       23
23       22
25       22
24       20
28       16
29       16
26       13
31       12
27       10
32        9
30        8
33        6
34        5
36        4
35        4
43        3
38        3
41        2
50        2
37        2
51        1
55        1
46        1
58        1
44        1
39        1
63        1
Name: campaign, dtype: int64

## Class Distribution

In [36]:
print("Originally encoded as: \n\n", df.y.unique())

Originally encoded as: 

 ['no' 'yes']


In [37]:
df["y"].replace(['no', 'yes'], 
                      [0, 1],
                      inplace = True)

In [38]:
df.y.dtype

dtype('int64')

In [39]:
label_pct = df['y'].value_counts(normalize = True)*100
label_ct =  df['y'].value_counts()
pd.DataFrame({'labels': label_pct.index, 'count': label_ct.values, 'percentage': label_pct.values})

labels  count  percentage
0       0  37104       92.76
1       1   2896        7.24

## Replacing 'unknown' values

In [40]:
df.replace('unknown', np.nan, inplace=True)
df.head()

age           job  marital  education  default  balance  housing  loan  \
0   58    management  married   tertiary        0     2143        1     0   
1   44    technician   single  secondary        0       29        1     0   
2   33  entrepreneur  married  secondary        0        2        1     1   
3   47   blue-collar  married        NaN        0     1506        1     0   
4   33           NaN   single        NaN        0        1        0     0   

  contact  day month  duration  campaign  y  
0     NaN    5   may       261         1  0  
1     NaN    5   may       151         1  0  
2     NaN    5   may        76         1  0  
3     NaN    5   may        92         1  0  
4     NaN    5   may       198         1  0

## Checking for missing values

In [41]:
# Compute percentage of missing values
missing_values = (df.isnull().sum()/df.shape[0])*100

# Dataframe of missing values
missing_df = pd.DataFrame(
    {
        'Variable': missing_values.index, 
        '% missing values': missing_values.values
    }
)
missing_df.sort_values(
    by = '% missing values', 
    ascending=False)\
.reset_index(drop=True)

Variable  % missing values
0     contact           31.9125
1   education            3.8275
2         job            0.5875
3         age            0.0000
4     marital            0.0000
5     default            0.0000
6     balance            0.0000
7     housing            0.0000
8        loan            0.0000
9         day            0.0000
10      month            0.0000
11   duration            0.0000
12   campaign            0.0000
13          y            0.0000

## Drop Contact

31.9% of data contains missing values, contact also doesnt provide valuable information so lets drop the contact column

In [42]:
df.drop("contact", 1, inplace = True)

## Drop missing values

In [43]:
df.dropna(inplace = True)

In [44]:
print('Data size',df.shape)
df.head()

Data size (38338, 13)


age           job  marital  education  default  balance  housing  loan  \
0   58    management  married   tertiary        0     2143        1     0   
1   44    technician   single  secondary        0       29        1     0   
2   33  entrepreneur  married  secondary        0        2        1     1   
5   35    management  married   tertiary        0      231        1     0   
6   28    management   single   tertiary        0      447        1     1   

   day month  duration  campaign  y  
0    5   may       261         1  0  
1    5   may       151         1  0  
2    5   may        76         1  0  
5    5   may       139         1  0  
6    5   may       217         1  0

In [45]:
# Compute percentage of missing values
missing_values = (df.isnull().sum()/df.shape[0])*100

# Dataframe of missing values
missing_df = pd.DataFrame(
    {
        'Variable': missing_values.index, 
        '% missing values': missing_values.values
    }
)
missing_df.sort_values(
    by = '% missing values', 
    ascending=False)\
.reset_index(drop=True)

Variable  % missing values
0         age               0.0
1         job               0.0
2     marital               0.0
3   education               0.0
4     default               0.0
5     balance               0.0
6     housing               0.0
7        loan               0.0
8         day               0.0
9       month               0.0
10   duration               0.0
11   campaign               0.0
12          y               0.0

## Save as csv dataframe

In [46]:
df.to_csv('../data/clean_data.csv', index=False)

## Summary

job, education, and contact contained 'unknown' value

Contact column contained 31% unknown, also doesnt provide useful information, thus the column is dropped 

All other unknown values were dropped 

default, housing, loan, and y values were encoded as ['yes', 'no'], these were changed to binary ['1', '0']

